### PR TITLE
Tag alpha/beta releases

### PR DESCRIPTION
**Pull request by sonalideshpandemsft**
This change adds a new flag `packageReleaseType` to `flub generate buildVersion` command which includes options as `alpha`, `beta`, `public`, `untrimmed` and `none`. 

Prerelease/dev releases are recognized when the keyword `dev` is included in the version. Likewise, we can distinguish alpha and beta prereleases through the use of the tags `dev-alpha-types` and `dev-beta-types` respectively.

Release: 'prerelease', Types: beta = 2.0.0-dev-beta-types.7.1.0.12345
Release: 'prerelease', Types: alpha = 2.0.0-dev-alpha-types.7.1.0.12345
Release: 'prerelease', Types: none = 2.0.0-dev.7.1.0.12345
Release: 'prerelease', Types: public = 2.0.0-dev.7.1.0.12345
Release: 'prerelease', Types: untrimmed = 2.0.0-dev.7.1.0.12345

### CODE REVIEW

**Comment by tylerbutler** [comment]
How is this going to be used in the CI pipeline?

**Comment by sonalideshpandemsft** [comment]
> How is this going to be used in the CI pipeline?

There is a separate PR for pipeline changes: https://github.com/microsoft/FluidFramework/pull/17460/files

**Comment by tylerbutler** [comment]
> > How is this going to be used in the CI pipeline?
> 
> There is a separate PR for pipeline changes: [#17460 (files)](https://github.com/microsoft/FluidFramework/pull/17460/files).
> 
> I see `flub generate buildVersion` used here: [`main`/tools/pipelines/templates/include-set-package-version.yml#L107](https://github.com/microsoft/FluidFramework/blob/main/tools/pipelines/templates/include-set-package-version.yml?rgh-link-date=2023-09-25T20%3A31%3A23Z#L107). Right now, it fails as there are no alpha/beta `Release Build` options.

So you're planning to change the versions as well as the types? If we're introducing a new version scheme then we need a lot of testing in the version-utils package to make sure everything works with versions of the new kind. We already have a lot of problems with the test branch versions, for example, so I don't want to add new version formats without a lot of testing.

I think you probably should separate "releasing an alpha/beta package" and "updating the default types in the pipeline". Think about experimental packages, for example. We may want to use alpha types for those packages, even when we're not doing an "alpha" release. How would we do that? One answer might be to add a build step to the experimental packages that sets their types in the build.

**Comment by sonalideshpandemsft** [comment]
Just to unblock the alpha/beta work and as you suggested separating the release alpha/beta, this PR should be enough: https://github.com/microsoft/FluidFramework/pull/17484

**Comment by sonalideshpandemsft** [comment]
General question: Since now, we are planning to include alpha/beta api releases, why don't we consider experimental packages as alpha/beta releases?

File: build-tools/packages/build-cli/src/commands/generate/buildVersion.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ export default class GenerateBuildVersionCommand extends BaseCommand<
 		}),
 		release: Flags.string({
 			description: "Indicates the build is a release build.",
-			options: ["release", "prerelease", "none"],
+			options: ["release", "prerelease", "prerelease - alpha", "prerelease - beta", "none"],
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/generate/buildVersion.ts:37]: I think the alpha/beta/public/untrimmed/default types setting should be separate from the release type. The "default" would be to publish whatever package.json says. The other options would set the types and update the version right before the pack stage in the pipeline.

File: build-tools/packages/version-tools/src/versions.ts [hunk 2]
```diff
@@ -33,7 +34,20 @@ export function getSimpleVersion(
 		}
 
 		if (!argRelease) {
-			version = changePreReleaseIdentifier(version, "dev");
+			let preReleaseIdentifier = "dev";
+
+			switch (true) {
+				case packageTypesField === "beta":
+					preReleaseIdentifier = "dev-beta-types";
```
Comment (sonalideshpandemsft) [inline build-tools/packages/version-tools/src/versions.ts:41]: should there be version bag for public releases? (similar to alpha as `dev-alpha-types`) I'm confused between public and untrimmed versioning.

**Comment by tylerbutler** [comment]
I'm still really confused about what the alpha/beta types versions are going to be. Can you clarify in the PR description? Something like:

Release: 'release', Types: alpha = `2.0.0-internal.7.1.0-alpha-types` or `2.0.0-alpha-types`
Release: 'prerelease', Types: beta = `2.0.0-dev.7.1.0.12345-beta-types` or `2.0.0-beta-types`
Release: 'none', Types: alpha = `2.0.0-dev.7.1.0.12345-alpha-types` or `2.0.0-alpha-types`

etc.

**Comment by sonalideshpandemsft** [comment]
> I'm still really confused about what the alpha/beta types versions are going to be. Can you clarify in the PR description? Something like:
> 
> Release: 'release', Types: alpha = `2.0.0-internal.7.1.0-alpha-types` or `2.0.0-alpha-types` Release: 'prerelease', Types: beta = `2.0.0-dev.7.1.0.12345-beta-types` or `2.0.0-beta-types` Release: 'none', Types: alpha = `2.0.0-dev.7.1.0.12345-alpha-types` or `2.0.0-alpha-types`
> 
> etc.

updated the description. alpha and beta version bag would be available only for prereleases.

**Comment by tylerbutler** [comment]
I've been thinking about this more, and I'm leaning towards not using `flub generate buildVersion` for this purpose. In my mind, changing the types should be a self-contained thing that is run on top of the existing pipeline. I'm really nervous about the version scheme we pick and making sure it doesn't conflict with anything else, and I'd feel better if we didn't change the core versioning logic of the build and instead run a separate command to generate the alpha/beta types version _from_ the version of the build, maybe just by appending the beta-types/alpha-types/etc. to the calculated version string. That seems easier than changing the prerelease identifier, but it doesn't matter that much.

This could live in version-tools instead of build-cli, and would take a version and a types value and return the correct version string. We will need to make sure that the alpha and beta schemes do not get included with our version ranges. I think you have some test-cases, but I think you'll need more. Maybe look at the versions of a package we publish to the test feed and the build feed and use those to build test cases. You'll also want to add test cases for any existing function that takes a version, testing that it works with the new format. Maybe some functions should throw, maybe some should consider the versions valid or invalid, etc. Adding test cases will help discover what the behavior is with the new versions so you can decide wther that behavior is OK or not.

On the code changes, as it is you have a new parameter to getSimpleVersion, which is weird because that's only used in a few pipelines. I think instead you want to either add some new functions that take versions and return alpha/beta ones, or update the "internal version scheme" functions to work with these, or something else. Adding test cases will really help. If the test cases are thorough, it's easy to tell by reviewing them whether the logic is correct or makes sense even without reading the actual code. So I would suggest starting with a list of test versions and write the test cases first.

Regardless of how the version is calculated, you'll need to set it explicitly in the pipeline and update deps. In the step where you update the types, you would also update the version using something like `flub bump client --exact 2.0.0-internal.7.1.0-alpha-types --interdependencyRange="" -x`.

> alpha and beta version bag would be available only for prereleases.

How are you enforcing this?

**Pull request closed without merging**